src/classes/Mfa.php

```diff
@@ -0,0 +1,199 @@
+<?php
+/**
+ * @package   Elabftw\Elabftw
+ * @author    Nicolas CARPi <nico-git@deltablot.email>
+ * @author    Marcel Bolten
+ * @copyright 2012 Nicolas CARPi
+ * @license   https://www.gnu.org/licenses/agpl-3.0.html AGPL-3.0
+ * @see       https://www.elabftw.net Official website
+ */
+declare(strict_types=1);
+
+namespace Elabftw\Elabftw;
+
+use Elabftw\Services\Filter;
+use Elabftw\Services\MpdfQrProvider;
+use RobThree\Auth\TwoFactorAuth;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Session;
+use function time;
+
+/**
+ * Provide methods for multi/two-factor authentication
+ */
+class Mfa
+{
+    /** @var Session $Session Current session */
+    public $Session;
+
+    /** @var Db $Db SQL Database */
+    protected $Db;
+
+    /** @var TwoFactorAuth $TwoFactorAuth PHP Class for handling two/multi-factor authentication */
+    private $TwoFactorAuth;
+
+    /** @var Request $Request Current request */
+    private $Request;
+
+    /**
+     * Constructor
+     *
+     * @param Request $request
+     * @param Session<mixed> $session
+     */
+    public function __construct(Request $request, Session $session)
+    {
+        $this->TwoFactorAuth = new TwoFactorAuth('eLabFTW', 6, 30, 'sha1', new MpdfQrProvider());
+        $this->Db = Db::getConnection();
+        $this->Request = $request;
+        $this->Session = $session;
+    }
+
+    /**
+     * Generate a new MFA secret
+     * Redirect to multi-factor code submission
+     *
+     * @param string $redirect Where user will be redirected to after code submission
+     * @return string location where MFA verification takes place
+     */
+    public function enable(string $redirect): string
+    {
+        // Need to request verification code to confirm user got secret and can authenticate in the future by MFA
+        $this->Session->set('mfa_secret', $this->TwoFactorAuth->createSecret(160));
+        $this->Session->set('enable_mfa', true);
+        $this->Session->set('mfa_redirect', $redirect);
+
+        return (string) '../../login.php';
+    }
+
+    /**
+     * Get QR code image with MFA secret as data URI
+     *
+     * @param string $email
+     * @return string
+     */
+    public function getQRCodeImageAsDataUri(string $email): string
+    {
+        return $this->TwoFactorAuth->getQRCodeImageAsDataUri($email, $this->Session->get('mfa_secret'));
+    }
+
+    /**
+     * Abort enable MFA
+     *
+     * @return string previously specified redirect location
+     */
+    public function abortEnable(): string
+    {
+        $this->Session->getFlashBag()->add('ko', _('Two Factor Authentication not enabled!'));
+        $location = $this->Session->get('mfa_redirect');
```

Comment:
`$location` is not used here. Found through: https://scrutinizer-ci.com/g/elabftw/elabftw/inspections/ab841c9a-62cd-4b66-84ce-ff1f13a3349d/issues/files/src/classes/Mfa.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

Reply:
This is a leftover. I removed it. MFA::cleanup() handles the redirect location now.